### PR TITLE
Add support for schema-based-sharding via a GUC

### PR DESCRIPTION
**Pull request by onurctirtir**
DESCRIPTION: Adds citus.enable_schema_based_sharding GUC that allows sharding the database based on schemas when enabled.

* Refactor the logic that automatically creates Citus managed tables 

* Refactor CreateSingleShardTable() to allow specifying colocation id instead

* Add support for schema-based-sharding via a GUC

### What this PR is about:
Add **citus.enable_schema_based_sharding GUC** to enable schema-based
sharding. Each schema created while this GUC is ON will be considered
as a tenant schema. Later on, regardless of whether the GUC is ON or
OFF, any table created in a tenant schema will be converted to a
single shard distributed table (without a shard key). All the tenant
tables that belong to a particular schema will be co-located with each
other and will have a shard count of 1.

We introduce a new metadata table --pg_dist_tenant_schema-- to do the
bookkeeping for tenant schemas:
```sql
psql> \d pg_dist_tenant_schema
          Table "pg_catalog.pg_dist_tenant_schema"
┌───────────────┬─────────┬───────────┬──────────┬─────────┐
│    Column     │  Type   │ Collation │ Nullable │ Default │
├───────────────┼─────────┼───────────┼──────────┼─────────┤
│ schemaid      │ oid     │           │ not null │         │
│ colocationid  │ integer │           │ not null │         │
└───────────────┴─────────┴───────────┴──────────┴─────────┘
Indexes:
    "pg_dist_tenant_schema_pkey" PRIMARY KEY, btree (schemaid)
    "pg_dist_tenant_schema_unique_colocationid_index" UNIQUE, btree (colocationid)

psql> table pg_dist_tenant_schema;
┌───────────┬───────────────┐
│ schemaid  │ colocationid  │
├───────────┼───────────────┤
│     41963 │            91 │
│     41962 │            90 │
└───────────┴───────────────┘
(2 rows)
```

Colocation id column of pg_dist_tenant_schema can never be NULL even
for the tenant schemas that don't have a tenant table yet. This is
because, we assign colocation ids to tenant schemas as soon as they
are created. That way, we can keep associating tenant schemas with
particular colocation groups even if all the tenant tables of a tenant
schema are dropped and recreated later on.

When a tenant schema is dropped, we delete the corresponding row from
pg_dist_tenant_schema. In that case, we delete the corresponding
colocation group from pg_dist_colocation as well.

### Future work for 12.0 release:
We're building schema-based sharding on top of the infrastructure that
adds support for creating distributed tables without a shard key (https://github.com/citusdata/citus/pull/6867).
However, not all the operations that can be done on distributed tables
without a shard key necessarily make sense (in the same way) in the
context of schema-based sharding. For example, we need to think about
what happens if user attempts altering schema of a tenant table. We
will tackle such scenarios in a future PR.

We will also add a new UDF --citus.schema_tenant_set() or such-- to
allow users to use an existing schema as a tenant schema, and another
one --citus.schema_tenant_unset() or such-- to stop using a schema as
a tenant schema in future PRs.

### CODE REVIEW

File: src/backend/distributed/commands/schema_based_sharding.c [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+	 * we want to treat "public" schema in the same way even if it's
+	 * recreated.
+	 */
+	if (strcmp(schemaName, "public") == 0)
+	{
+		return false;
+	}
+
+	/* TODO: Can it be a system or temp schema? does postgres ever use CREATE SCHEMA for them? */
+	/* TODO: Should we check whether the schema owned by an extension? */
```
Comment (onurctirtir) [inline src/backend/distributed/commands/schema_based_sharding.c:53]: .

File: src/include/distributed/commands.h [hunk 2]
```diff
@@ -24,6 +24,12 @@
 
 extern bool AddAllLocalTablesToMetadata;
 
+/*
+ * TODO: should we require AddAllLocalTablesToMetadata being false
+ *       to enable this?
+ */
```
Comment (onurctirtir) [inline src/include/distributed/commands.h:30]: .

File: src/backend/distributed/commands/schema_based_sharding.c
Comment (onurctirtir) [inline src/backend/distributed/commands/schema_based_sharding.c:1]: Should we automatically convert tables created in public schema to reference tables?

File: src/include/distributed/commands.h (hunk 2 above)
Comment (onurctirtir) [reply src/include/distributed/commands.h:30]: Maybe no, but we should check for schema-based sharding logic before citus managed tables branch (if / else if) to avoid going through table conversions twice.

File: src/backend/distributed/commands/schema_based_sharding.c
Comment (onurctirtir) [reply src/backend/distributed/commands/schema_based_sharding.c:1]: No

**Comment by onurctirtir** [comment]
#6871 will help fixing what flaky test detector complains about.

File: src/backend/distributed/metadata/metadata_sync.c [hunk 3]
```diff
@@ -4486,6 +4659,68 @@ SendColocationMetadataCommands(MetadataSyncContext *context)
...
+		StringInfo insertTenantSchemaCommand = makeStringInfo();
+		appendStringInfo(insertTenantSchemaCommand,
+						 "SELECT pg_catalog.citus_internal_insert_tenant_schema(%s, %u)",
+						 RemoteSchemaIdExpression(tenantSchemaForm->schemaid),
+						 colocationId);
+
+		List *commandList = list_make1(insertTenantSchemaCommand->data);
+		SendOrCollectCommandListToActivatedNodes(context, commandList);
+	}
+	MemoryContextSwitchTo(oldContext);
```
Comment (onurctirtir) [inline src/backend/distributed/metadata/metadata_sync.c:4717]: @aykut-bozkurt, I'd like to hear whether I'm using the new metadata-sync APIs correctly, to sync the entries of that new metadata table.

Comment (aykut-bozkurt) [reply src/backend/distributed/metadata/metadata_sync.c:4717]: Looks right.

File: src/test/regress/sql/schema_based_sharding.sql
Comment (onurctirtir) [inline src/test/regress/sql/schema_based_sharding.sql:1]: Need to find out what's happening:

```sql
SELECT citus_add_node('localhost', 9700, 0);
set citus.enable_schema_based_sharding to on;
create schema t10;
SELECT citus_add_node('localhost', 9701);
WARNING:  schema "t10" does not exist
ERROR:  failure on connection marked as essential: localhost:9701
```

Comment (onurctirtir) [reply src/test/regress/sql/schema_based_sharding.sql:1]: This is probably because the schema is empty, but wondering why this is not the case in regression tests.

File: src/test/regress/expected/multi_extension.out [hunk 4]
```diff
@@ -1762,6 +1780,46 @@ SELECT citus_remove_node('localhost', :master_port);
...
+(1 row)
+
+CREATE SCHEMA tenant_schema;
+-- Make sure that we can sync metadata for empty tenant schemas
+-- when adding the first node to the cluster.
+-- FIXME: Fix this before merging the PR.
+SELECT citus_add_node('localhost', :worker_1_port);
+WARNING:  schema "tenant_schema" does not exist
+CONTEXT:  while executing command on localhost:xxxxx
+ERROR:  failure on connection marked as essential: localhost:xxxxx
```
Comment (onurctirtir) [inline src/test/regress/expected/multi_extension.out:1820]: need to fix that

Comment (onurctirtir) [inline src/test/regress/expected/multi_extension.out:1820]: need to fix that

**Comment by onurctirtir** [comment]
Test failures are because of 13311adc53e7debfc84a161b725c94ed1a8d405a. I'll fix those in a separate PR (#6900) and then rebased this one onto main then.

File: src/backend/distributed/metadata/metadata_sync.c [hunk 5]
```diff
@@ -170,6 +173,9 @@ PG_FUNCTION_INFO_V1(citus_internal_update_relation_colocation);
 PG_FUNCTION_INFO_V1(citus_internal_add_object_metadata);
 PG_FUNCTION_INFO_V1(citus_internal_add_colocation_metadata);
 PG_FUNCTION_INFO_V1(citus_internal_delete_colocation_metadata);
+PG_FUNCTION_INFO_V1(citus_internal_insert_tenant_schema);
```
Comment (marcocitus) [inline src/backend/distributed/metadata/metadata_sync.c:176]: so far we used "add" instead of "insert"

File: src/backend/distributed/commands/schema_based_sharding.c [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+
+	CreateSingleShardTable(relationId, colocateWithTableName);
+
+	if (firstTableInSchema)
+	{
+		/*
+		 * Save it into pg_dist_tenant_schema if this is the first tenant
+		 * table in the schema.
+		 */
+		SetTenantSchemaColocationId(schemaId, TableColocationId(relationId));
```
Comment (marcocitus) [inline src/backend/distributed/commands/schema_based_sharding.c:172]: I have some concerns about the approach of repeatedly resetting and recreating the colocation ID when a schema goes empty. As a user, I expect a strong coupling between the schema and my tenant, but this effectively creates a new tenant for the purpose of citus_stat_tenants and future tenant pinning. Is there a way to preserve the co-location ID of empty schemas?

File: src/backend/distributed/shared_library_init.c [hunk 7]
```diff
@@ -1374,6 +1374,20 @@ RegisterCitusConfigVariables(void)
 		GUC_NO_SHOW_ALL,
 		NULL, NULL, NULL);
 
+	DefineCustomBoolVariable(
+		"citus.enable_schema_based_sharding",
```
Comment (marcocitus) [inline src/backend/distributed/shared_library_init.c:1378]: I'm kind of wondering whether this should be schema_based_sharding or schema_sharding.

I do feel like schema_based_sharding sounds a bit better, but if we ever add database sharding, then database_based_sharding sounds much worse than database_sharding. :)

File: src/backend/distributed/sql/citus--11.3-1--12.0-1.sql [hunk 8]
```diff
@@ -1,3 +1,21 @@
 -- citus--11.3-1--12.0-1
 
 -- bump version to 12.0-1
+
+CREATE TABLE citus.pg_dist_tenant_schema (
+    schema_id oid NOT NULL,
```
Comment (onurctirtir) [inline src/backend/distributed/sql/citus--11.3-1--12.0-1.sql:6]: use regnamespace instead?

Comment (marcocitus) [reply src/backend/distributed/sql/citus--11.3-1--12.0-1.sql:6]: no, that breaks pg_upgrade https://www.postgresql.org/docs/current/pgupgrade.html

File: src/include/distributed/metadata_utility.h [hunk 9]
```diff
@@ -201,6 +201,37 @@ typedef enum SizeQueryType
 	TABLE_SIZE /* pg_table_size() */
 } SizeQueryType;
 
+
+typedef enum
+{
+	COLOCATE_USING_TABLE_LIKE_OPT,
```
Comment (marcocitus) [inline src/include/distributed/metadata_utility.h:207]: COLOCATE_WITH?

File: src/backend/distributed/sql/citus--11.3-1--12.0-1.sql (hunk 8 above)
Comment (onurctirtir) [reply src/backend/distributed/sql/citus--11.3-1--12.0-1.sql:6]: > are schema OIDs stable across pg_upgrade?

I don't think so.

Comment (onurctirtir) [reply src/backend/distributed/sql/citus--11.3-1--12.0-1.sql:6]: > no, that breaks pg_upgrade https://www.postgresql.org/docs/current/pgupgrade.html

Given that we anyway need to save & reload pg_dist_tenant_schema during pg-upgrades (because oids are not stable), would it still be bad to use regnamespace?

File: src/backend/distributed/commands/schema.c [hunk 10]
```diff
@@ -74,6 +76,19 @@ PreprocessCreateSchemaStmt(Node *node, const char *queryString,
 
 	commands = list_concat(commands, GetGrantCommandsFromCreateSchemaStmt(node));
 
+	if (ShouldUseSchemaAsTenantSchema(createSchemaStmt->schemaname))
```
Comment (marcocitus) [inline src/backend/distributed/commands/schema.c:79]: name is a bit hard to follow, maybe ShouldUseSchemaSharding / ShouldUseSchemaBasedSharding?

File: src/backend/distributed/sql/citus--11.3-1--12.0-1.sql (hunk 8 above)
Comment (onurctirtir) [reply src/backend/distributed/sql/citus--11.3-1--12.0-1.sql]: I cannot see any Postgres catalog tables that has a regnamespace column, so it might be nicer to keep using Oid.

File: src/backend/distributed/commands/table.c [hunk 11]
```diff
@@ -282,6 +282,15 @@ PostprocessCreateTableStmtForeignKeys(CreateStmt *createStatement)
 	bool missingOk = false;
 	Oid relationId = RangeVarGetRelid(createStatement->relation, NoLock, missingOk);
 
+	if (ShouldCreateTenantTable(relationId))
+	{
+		/*
+		 * Avoid unnecessarily adding the table into metadata if we will
```
Comment (marcocitus) [inline src/backend/distributed/commands/table.c:288]: I kind of wonder whether this function should deal with CreateStmt more broadly.

It's a bit weird that we have auto-convert-on-CREATE-TABLE logic dispersed across different files

File: src/backend/distributed/commands/schema_based_sharding.c [hunk 12]
```diff
@@ -0,0 +1,221 @@
...
+		return false;
+	}
+
+	return true;
+}
+
+
+/*
+ * ShouldCreateTenantTable returns true if given parsetree is a CREATE TABLE
+ * statement and the table should be treated as a tenant table.
```
Comment (marcocitus) [inline src/backend/distributed/commands/schema_based_sharding.c:81]: comment does not match the function declarawtion

File: src/include/distributed/metadata_utility.h [hunk 13]
```diff
@@ -201,6 +201,37 @@ typedef enum SizeQueryType
...
+ * Among the functions used to create a Citus table, right now only
+ * CreateSingleShardTable() accepts a ColocationParam.
+ */
+typedef struct
+{
+	union
+	{
+		char *colocateWithTableName;
+		uint32 colocationId;
+	};
```
Comment (marcocitus) [inline src/include/distributed/metadata_utility.h:229]: if it simplifies the syntax, we can afford to have both fields in top-level struct

File: src/backend/distributed/commands/schema.c [hunk 14]
```diff
@@ -109,6 +123,17 @@ PreprocessDropSchemaStmt(Node *node, const char *queryString,
 
 	String *schemaVal = NULL;
 	foreach_ptr(schemaVal, distributedSchemas)
+	{
+		bool missingOk = false;
+		Oid schemaId = get_namespace_oid(strVal(schemaVal), missingOk);
+		if (IsTenantSchema(schemaId))
+		{
+			UnregisterTenantSchema(schemaId);
```
Comment (onurctirtir) [inline src/backend/distributed/commands/schema.c:131]: We need to delete pg_dist_colocation entry too. If the schema has at least one tenant table, we already do so via master_remove_partition_metadata(). However, this is not the case for an empty schema.

File: src/backend/distributed/executor/adaptive_executor.c [hunk 15]
```diff
@@ -1040,6 +1040,12 @@ ExecuteTaskListIntoTupleDest(RowModifyLevel modLevel, List *taskList,
 uint64
 ExecuteTaskListExtended(ExecutionParams *executionParams)
 {
+	/* if there are no tasks to execute, we can return early */
```
Comment (aykut-bozkurt) [inline src/backend/distributed/executor/adaptive_executor.c:1043]: Why do we need that? Is that needed by this PR?

File: src/backend/distributed/commands/utility_hook.c [hunk 16]
```diff
@@ -344,26 +344,52 @@ multi_ProcessUtility(PlannedStmt *pstmt,
...
 				 * Here we set autoConverted to false, since the user explicitly
 				 * wants these tables to be added to metadata, by setting the
 				 * GUC use_citus_managed_tables to true.
 				 */
 				bool autoConverted = false;
 				bool cascade = true;
-				CreateCitusLocalTable(relationId, cascade, autoConverted);
+				CreateCitusLocalTable(createdRelationId, cascade, autoConverted);
 			}
 		}
```
Comment (aykut-bozkurt) [inline src/backend/distributed/commands/utility_hook.c:394]: Can we move this logic into a function if possible? To make `multi_processutility` function, which is already long enough, more readable.

**Review by aykut-bozkurt — COMMENTED** [review]
Looks great overall.! Just added few minors.

File: src/backend/distributed/executor/adaptive_executor.c (hunk 15 above)
Comment (onurctirtir) [reply src/backend/distributed/executor/adaptive_executor.c:1043]: This change is made in a commit 6c0e3cbcf91af00185cd0b63c6a988de6d7ca218, which I opened a separate PR for (https://github.com/citusdata/citus/pull/6900). I included that commit in that PR too to make the tests passing here as well.

File: src/backend/distributed/commands/utility_hook.c (hunk 16 above)
Comment (onurctirtir) [reply src/backend/distributed/commands/utility_hook.c:394]: For such slightly complex logics, I feel like doing the things explicitly is better than hiding the details. But ofc I'm open to suggestions too, any ideas about how to improve there?

Comment (marcocitus) [reply src/backend/distributed/commands/utility_hook.c:394]: seems like we could do something like:
```c
if (context == PROCESS_UTILITY_TOPLEVEL &&
	(IsA(parsetree, CreateStmt) || IsA(parsetree, CreateForeignTableStmt)))
{
    ConvertNewTableIfNecessary(parsetree);
}
```

File: src/backend/distributed/commands/schema.c (hunk 14 above)
Comment (onurctirtir) [reply src/backend/distributed/commands/schema.c:131]: oh, I don't think so

**Review by marcocitus — COMMENTED** [review]
we

File: src/backend/distributed/commands/utility_hook.c [hunk 17]
```diff
@@ -344,26 +343,19 @@ multi_ProcessUtility(PlannedStmt *pstmt,
...
-				 * Here we need to increment command counter so that next command
-				 * can see the new table.
-				 */
-				CommandCounterIncrement();
-				CreateStmt *createTableStmt = (CreateStmt *) parsetree;
-				Oid relationId = RangeVarGetRelid(createTableStmt->relation,
-												  NoLock, false);
+				CreateStmt *baseCreateTableStmt =
+					IsA(parsetree, CreateStmt) ? (CreateStmt *) parsetree :
+					(CreateStmt *) &(((CreateForeignTableStmt *) parsetree)->base);
```
Comment (marcocitus) [inline src/backend/distributed/commands/utility_hook.c:356]: maybe add a comment on why this is not just casting to CreateStmt

(my guess is that this protects us from future postgres changes?)

File: src/test/regress/sql/schema_based_sharding.sql [hunk 18]
```diff
@@ -0,0 +1,412 @@
...
+
+-- Verify that we don't unnecessarily convert a table into a Citus managed
+-- table when creating it with a pre-defined foreign key to a reference table.
+CREATE TABLE reference_table(a int PRIMARY KEY);
+SELECT create_reference_table('reference_table');
+
+-- Notice that tenant_7.tbl_4 have foreign keys both to tenant_7.tbl_3 and
+-- to reference_table.
+CREATE TABLE tenant_7.tbl_4(a int REFERENCES reference_table, FOREIGN KEY(a) REFERENCES tenant_7.tbl_3(a));
+
```
Comment (marcocitus) [inline src/test/regress/sql/schema_based_sharding.sql:342]: could be nice to test ON DELETE .. CASCADE behaviour

File: src/test/regress/sql/schema_based_sharding.sql [hunk 19]
```diff
@@ -0,0 +1,412 @@
...
+\c - - - :master_port
+
+SET search_path TO regular_schema;
+SET citus.next_shard_id TO 1940000;
+SET citus.shard_count TO 32;
+SET citus.shard_replication_factor TO 1;
+SET client_min_messages TO NOTICE;
+
+CREATE TABLE tenant_3.tbl_1(a int, b text);
```
Comment (marcocitus) [inline src/test/regress/sql/schema_based_sharding.sql:386]: would be good to have a few search_path switching tests, especially with prepared statements and maybe functions, since that's a pattern that's quite unique to schema-based sharding.

File: src/backend/distributed/sql/citus--11.3-1--12.0-1.sql [hunk 20]
```diff
@@ -1,3 +1,26 @@
...
+-- udfs used to modify pg_dist_tenant_schema on workers, to sync metadata
+#include "udfs/citus_internal_add_tenant_schema/12.0-1.sql"
+#include "udfs/citus_internal_delete_tenant_schema/12.0-1.sql"
+
+#include "udfs/citus_prepare_pg_upgrade/12.0-1.sql"
+#include "udfs/citus_finish_pg_upgrade/12.0-1.sql"
+
+-- udfs used to modify pg_dist_tenant_schema globally via drop trigger
+#include "udfs/citus_internal_delete_colocation_metadata_globally/12.0-1.sql"
+#include "udfs/citus_internal_delete_tenant_schema_globally/12.0-1.sql"
```
Comment (onurctirtir) [inline src/backend/distributed/sql/citus--11.3-1--12.0-1.sql:25]: Add a few simple tests for those two too.

File: src/test/regress/sql/schema_based_sharding.sql
Comment (onurctirtir) [inline src/test/regress/sql/schema_based_sharding.sql:1]: Add tests where a non-super user creates tenant schemas / tenant schema tables:

```sql
create role test_non_super_user;
alter role test_non_super_user with login;
grant create on database onurctirtir to test_non_super_user;
SELECT run_command_on_workers('grant create on database onurctirtir to test_non_super_user ');
```

```sql
\c - test_non_super_user

SET citus.enable_schema_based_sharding TO ON;

create schema tenant_schema;
...
NOTICE: issuing SELECT pg_catalog.citus_internal_add_tenant_schema('t3'::regnamespace, 3)
DETAIL: on server test_non_super_user@localhost:9702 connectionId: 4
...
ERROR:  operation is not allowed
HINT:  Run the command with a superuser.
CONTEXT:  while executing command on localhost:9702
```

File: src/backend/distributed/metadata/metadata_sync.c [hunk 21]
```diff
@@ -3937,6 +3999,92 @@ ColocationGroupDeleteCommand(uint32 colocationId)
...
+}
+
+
+/*
+ * TenantSchemaDeleteCommandBySchemaId returns a command to call
+ * citus_internal_delete_tenant_schema() based on fiven given
+ * schemaId.
+ */
+char *
+TenantSchemaDeleteCommandBySchemaId(Oid schemaId)
```
Comment (marcocitus) [inline src/backend/distributed/metadata/metadata_sync.c:4024]: seems unused

File: src/backend/distributed/sql/udfs/citus_drop_trigger/latest.sql [hunk 22]
```diff
@@ -34,6 +35,25 @@ BEGIN
...
 
+    -- Remove entries from pg_catalog.pg_dist_tenant_schema for all dropped tenant schemas.
+    -- Also delete the corresponding colocation group from pg_catalog.pg_dist_colocation.
+    --
+    -- Although normally we automatically delete the colocation groups when they become empty,
+    -- we don't do so for the colocation groups that are created for tenant schemas. For this
+    -- reason, here we need to delete the colocation group when the tenant schema is dropped.
+    FOR v_obj IN SELECT * FROM pg_event_trigger_dropped_objects() WHERE object_type = 'schema'
+    LOOP
+        IF EXISTS (SELECT 1 FROM pg_catalog.pg_dist_tenant_schema WHERE schemaid = v_obj.objid)
```
Comment (marcocitus) [inline src/backend/distributed/sql/udfs/citus_drop_trigger/latest.sql:46]: this could be a join in the query above

Comment (marcocitus) [reply src/backend/distributed/sql/udfs/citus_drop_trigger/latest.sql:46]: (alternatively, could be folded into loop below)

File: src/backend/distributed/sql/udfs/citus_drop_trigger/latest.sql [hunk 23]
```diff
@@ -34,6 +35,25 @@ BEGIN
...
+        THEN
+            tenant_schema_colocation_id := (
+                SELECT colocationid FROM pg_catalog.pg_dist_tenant_schema WHERE schemaid = v_obj.objid
+            );
+            PERFORM citus_internal_delete_colocation_metadata_globally(tenant_schema_colocation_id);
+
+            PERFORM citus_internal_delete_tenant_schema_globally(v_obj.objid, v_obj.object_name);
```
[the remark this answers is not in the document]
Comment (onurctirtir) [reply src/backend/distributed/sql/udfs/citus_drop_trigger/latest.sql]: > can't we just make this a local-only operation that runs when the local schema is dropped?

Done now.

File: src/backend/distributed/sql/udfs/citus_drop_trigger/latest.sql (hunk 22 above)
Comment (onurctirtir) [reply src/backend/distributed/sql/udfs/citus_drop_trigger/latest.sql:45]: (will do soon 👍)

File: src/backend/distributed/commands/table.c [hunk 24]
```diff
@@ -282,6 +282,15 @@ PostprocessCreateTableStmtForeignKeys(CreateStmt *createStatement)
 	bool missingOk = false;
 	Oid relationId = RangeVarGetRelid(createStatement->relation, NoLock, missingOk);
 
+	if (ShouldCreateTenantSchemaTable(relationId))
+	{
+		/*
+		 * Avoid unnecessarily adding the table into metadata if we will
+		 * distribute it as a tenant table later.
+		 */
+		return;
```
Comment (marcocitus) [inline src/backend/distributed/commands/table.c:291]: still kind of wondering whether we can call ConvertNewTableIfNecessary from here

File: src/backend/distributed/commands/utility_hook.c (hunk 17 above)
Comment (marcocitus) [reply src/backend/distributed/commands/utility_hook.c:356]: bump, why is this not just `CreateStmt *baseCreateTableStmt = (CreateStmt *) parsetree;` ? could use a comment

File: src/backend/distributed/utils/colocation_utils.c [hunk 25]
```diff
@@ -1422,4 +1454,24 @@ EnsureTableCanBeColocatedWith(Oid relationId, char replicationModel,
...
+		char *sourceRelationName = get_rel_name(sourceRelationId);
+		char *sourceRelationSchemaName = get_namespace_name(sourceRelationSchemaId);
+
+		ereport(ERROR, (errmsg("cannot colocate tables %s and %s",
+							   sourceRelationName, relationName),
+						errdetail("Cannot colocate tables with tenant tables "
+								  "by using colocate_with option."),
+						errhint("Consider using \"ALTER TABLE %s SET SCHEMA %s;\" "
+								"to move the table to the same schema with the "
+								"tenant table.",
```
Comment (marcocitus) [inline src/backend/distributed/utils/colocation_utils.c:1474]: is that possible?

**Review by marcocitus — COMMENTED** [review]
Looks pretty good overall, after addressing open comments this seems ready to merge.

File: src/backend/distributed/commands/utility_hook.c (hunk 17 above)
Comment (onurctirtir) [reply src/backend/distributed/commands/utility_hook.c:356]: > (my guess is that this protects us from future postgres changes?)

true, adding a comment for that

File: src/backend/distributed/commands/table.c (hunk 24 above)
Comment (onurctirtir) [reply src/backend/distributed/commands/table.c:291]: Hmm maybe we can do but would might mean a somewhat big change in the context of this PR --need to understand why we didn't do so for ConvertToManagedTable before.

File: src/backend/distributed/utils/colocation_utils.c (hunk 25 above)
Comment (onurctirtir) [reply src/backend/distributed/utils/colocation_utils.c:1474]: This is not possible today. But a time ago, I was thinking that we would definitely have such a logic for ALTER TABLE SET SCHEMA, in 12.0. Let me drop the hint and then we can re-add it if we decide implementing it in this release.

File: src/backend/distributed/sql/udfs/citus_drop_trigger/latest.sql (hunk 22 above)
Comment (onurctirtir) [reply src/backend/distributed/sql/udfs/citus_drop_trigger/latest.sql]: done

**Review by marcocitus — COMMENTED** [review]
We probably need to do something about this syntax, since it currently results in a local table existing in a tenant schema:
```
create schema t2 create table test (x int, y int);
```

**Comment by aykut-bozkurt** [comment]
nit about weird syntax:

```sql
CREATE SCHEMA "CiTuS.TeeN";
```
This fails like below:
```sql
NOTICE:  issuing SELECT pg_catalog.citus_internal_add_tenant_schema("CiTuS.TeeN"::regnamespace, 1)
DETAIL:  on server aykutbozkurt@localhost:9701 connectionId: 1
ERROR:  column "CiTuS.TeeN" does not exist
```

We can use `quote_literal_cstr(quote_identifier(schemaName))` at `RemoteSchemaIdExpression` to fix.

File: src/test/regress/sql/schema_based_sharding.sql [hunk 26]
```diff
@@ -0,0 +1,700 @@
...
+SET citus.shard_replication_factor TO 1;
+SET client_min_messages TO NOTICE;
+SET citus.enable_schema_based_sharding TO ON;
+
+-- test create / drop tenant schema / table
+
+CREATE SCHEMA tenant_10;
+CREATE TABLE tenant_10.tbl_1(a int, b text);
+CREATE TABLE tenant_10.tbl_2(a int, b text);
+
```
Comment (marcocitus) [inline src/test/regress/sql/schema_based_sharding.sql:561]: consider adding a `CREATE TABLE tenant_10.tbl_3 AS SELECT s, s FROM generate_series(1,100) s;` test

This currently results in a local table being created in the schema.

**Comment by onurctirtir** [comment]
> I had 2 questions. First one looks like works and is resolved. The second one is a problem. We can error and defer it to the second phase or see the solution I suggest.
> 
> 1. What we do when we alter a tenant schema? Looks like it just works we do not hardcode any schema's name into any table. Verified with below.
> 
> ```sql
> ALTER SCHEMA tenant1 RENAME TO <NEW_SCHEMA_NAME>;
> ```
> 
> 2. What happens if we alter a tenant table's schema? Looks like queries still work and table colocation and other metadata are fine. But feels like bad to have the tables in different schemas with the same colocation entry.
> 
> ```sql
> ALTER TABLE tenant1.test1 SET SCHEMA <NEW_SCHEMA_NAME>;
> ```
> 
> * we may delete metadata if we change the tenant schema to a regular schema. (undistribute or leave it single shard?)
> * we may update metadata and move the single shard table if we change the tenant schema to another tenant schema. (This can be useful if user chooses a tenant schema by mistake)

Scenarios like 2) will be tackled in a separate PR.

File: src/test/regress/sql/schema_based_sharding.sql (hunk 26 above)
Comment (onurctirtir) [reply src/test/regress/sql/schema_based_sharding.sql:802]: now we're throwing an error for that

**Review by marcocitus — APPROVED** [review]
